[JSC] Add B3::MemoryCopy and B3::MemoryFill

Constellation opened the pull request:
#### 229760706c12a975b3f07f98c2c53a90abcbb42c
<pre>
[JSC] Add B3::MemoryCopy and B3::MemoryFill
<a href="https://bugs.webkit.org/show_bug.cgi?id=298906">https://bugs.webkit.org/show_bug.cgi?id=298906</a>
<a href="https://rdar.apple.com/160649162">rdar://160649162</a>

Reviewed by Yijia Huang.

This patch adds B3 MemoryCopy and MemoryFill values. They reserve
memcpy / memset semantics, and B3 strength reduction attempts to lower
them to sequence of loads and stores for small constant size. If the
size is unknown, it gets lowered to a function call eventually.
By using this, we refresh Wasm MemoryCopy and MemoryFill
implementations. They do bound check and emit these B3 values. And if
possible, they get lowered to loads / stores, omitting function calls.
This is useful since we are finding many wasm bulk memory.copy /
memory.fill are getting constant count via inlining and we can avoid
calling a function for these cases.

Tests: Source/JavaScriptCore/b3/testb3_1.cpp
       Source/JavaScriptCore/b3/testb3_8.cpp
* Source/JavaScriptCore/JavaScriptCore.xcodeproj/project.pbxproj:
* Source/JavaScriptCore/Sources.txt:
* Source/JavaScriptCore/b3/B3BulkMemoryValue.cpp: Added.
(JSC::B3::BulkMemoryValue::dumpMeta const):
(JSC::B3::BulkMemoryValue::BulkMemoryValue):
* Source/JavaScriptCore/b3/B3BulkMemoryValue.h: Added.
* Source/JavaScriptCore/b3/B3LowerInt64.cpp:
* Source/JavaScriptCore/b3/B3LowerMacrosAfterOptimizations.cpp:
* Source/JavaScriptCore/b3/B3LowerToAir.cpp:
* Source/JavaScriptCore/b3/B3LowerToAir32_64.cpp:
* Source/JavaScriptCore/b3/B3Opcode.h:
* Source/JavaScriptCore/b3/B3Operations.cpp: Added.
(JSC::B3::JSC_DEFINE_NOEXCEPT_JIT_OPERATION):
* Source/JavaScriptCore/b3/B3Operations.h: Added.
* Source/JavaScriptCore/b3/B3ReduceStrength.cpp:
* Source/JavaScriptCore/b3/B3Validate.cpp:
* Source/JavaScriptCore/b3/B3Value.cpp:
(JSC::B3::Value::effects const):
* Source/JavaScriptCore/b3/B3Value.h:
* Source/JavaScriptCore/b3/B3ValueInlines.h:
* Source/JavaScriptCore/b3/testb3.h:
* Source/JavaScriptCore/b3/testb3_1.cpp:
(run):
* Source/JavaScriptCore/b3/testb3_8.cpp:
(testMemoryCopy):
(testMemoryCopyConstant):
(testMemoryFill):
(testMemoryFillConstant):
* Source/JavaScriptCore/wasm/WasmOMGIRGenerator.cpp:
(JSC::Wasm::OMGIRGenerator::addMemoryFill):
(JSC::Wasm::OMGIRGenerator::addMemoryCopy):

Canonical link: <a href="https://commits.webkit.org/300210@main">https://commits.webkit.org/300210@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/4e372ac3bb30bff16020daa29567088644f1bb40

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows | Apple Internal |
| ----- | ---------------------- | ------- |  ----- |  --------- | ------ |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/121427 "2 style errors") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/41124 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/31783 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/127868 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/73510 "Built successfully") | [✅ 🛠 ios-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/4d7d8414-cf4f-4274-99cf-f0a86136d47b) 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/41826 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/49703 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/92235 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/61367 "Passed tests") | [✅ 🛠 mac-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/df80d4ce-aa89-46ad-a15f-1c04d110d0c5) 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/124379 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/33392 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/108786 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/72911 "Passed tests") | | [✅ 🛠 vision-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/5b3cc24a-46f7-4d8b-a2d1-23c7d40e70ac) 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/32408 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/26939 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/71448 "Built successfully") | | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/113557 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/102888 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/27113 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/130702 "Built successfully") | | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/119947 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/48355 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/36767 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/100825 "Passed tests") | | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/48723 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/104998 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/100732 "Passed tests") | | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/46144 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/24211 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/45051 "Built successfully") | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/19281 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/48213 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/53926 "Built successfully") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/150109 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/47685 "Built successfully") | | [❌ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/38193 "Found 1 new JSC stress test failure: stress/dont-link-virtual-calls-on-compiler-thread.js.no-llint (failure)") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/51031 "Built successfully") | | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/49367 "Built successfully") | | | | 
<!--EWS-Status-Bubble-End-->

justinmichaud commented on Source/JavaScriptCore/b3/B3Operations.cpp:
```diff
@@ -0,0 +1,50 @@
+
+JSC_DEFINE_NOEXCEPT_JIT_OPERATION(operationMemoryCopy, void, (void* dst, const void* src, size_t count))
+{
+    memmove(dst, src, count);
```
I have a question about this. I thought that the b3 memory model guaranteed that 64-bit writes wouldn't be torn. There isn't a safepoint in here, so I can see why this isn't going to crash. Similarly, it seems this is correct for shared arrays too (https://github.com/WebAssembly/threads/issues/9#issuecomment-725389629), so maybe I am imagining things.

Constellation replied:
This is aligned to Wasm MemoryCopy / MemoryFill semantics which is only applied to wasm memory. So basically no behavior change.

hyjorc1 commented on Source/JavaScriptCore/b3/B3ReduceStrength.cpp:
```diff
@@ -2310,6 +2312,147 @@ class ReduceStrength {
+        case MemoryFill: {
+            if (m_value->child(1)->hasInt() && m_value->child(2)->hasInt()) {
+                auto* memoryFill = m_value->as<BulkMemoryValue>();
+                uint64_t target = static_cast<uint8_t>(m_value->child(1)->asInt());
```
`static_cast<uint8_t>` is that correct?

hyjorc1 replied:
oh, nvm, memory.fill instruction takes an i32 value but only uses the lowest 8 bits as the fill byte.

hyjorc1 commented on Source/JavaScriptCore/b3/B3Validate.cpp:
```diff
@@ -428,6 +428,22 @@ class Validater {
+                VALIDATE(!value->kind().isChill(), ("At ", *value));
+                VALIDATE(value->numChildren() == 3, ("At ", *value));
+                VALIDATE(value->child(0)->type() == pointerType(), ("At ", *value));
+                VALIDATE(value->child(1)->type() == Int32, ("At ", *value));
```
Is `Int32` correct type for the 2nd arg of `MemoryFill`?

hyjorc1 replied:
oh, nvm, memory.fill instruction takes an i32 value but only uses the lowest 8 bits as the fill byte.

hyjorc1 reviewed (approved):
r=me

Constellation committed:
[JSC] Add B3::MemoryCopy and B3::MemoryFill
https://bugs.webkit.org/show_bug.cgi?id=298906
rdar://160649162

Reviewed by Yijia Huang.

This patch adds B3 MemoryCopy and MemoryFill values. They reserve
memcpy / memset semantics, and B3 strength reduction attempts to lower
them to sequence of loads and stores for small constant size. If the
size is unknown, it gets lowered to a function call eventually.
By using this, we refresh Wasm MemoryCopy and MemoryFill
implementations. They do bound check and emit these B3 values. And if
possible, they get lowered to loads / stores, omitting function calls.
This is useful since we are finding many wasm bulk memory.copy /
memory.fill are getting constant count via inlining and we can avoid
calling a function for these cases.

Tests: Source/JavaScriptCore/b3/testb3_1.cpp
       Source/JavaScriptCore/b3/testb3_8.cpp
* Source/JavaScriptCore/JavaScriptCore.xcodeproj/project.pbxproj:
* Source/JavaScriptCore/Sources.txt:
* Source/JavaScriptCore/b3/B3BulkMemoryValue.cpp: Added.
(JSC::B3::BulkMemoryValue::dumpMeta const):
(JSC::B3::BulkMemoryValue::BulkMemoryValue):
* Source/JavaScriptCore/b3/B3BulkMemoryValue.h: Added.
* Source/JavaScriptCore/b3/B3LowerInt64.cpp:
* Source/JavaScriptCore/b3/B3LowerMacrosAfterOptimizations.cpp:
* Source/JavaScriptCore/b3/B3LowerToAir.cpp:
* Source/JavaScriptCore/b3/B3LowerToAir32_64.cpp:
* Source/JavaScriptCore/b3/B3Opcode.h:
* Source/JavaScriptCore/b3/B3Operations.cpp: Added.
(JSC::B3::JSC_DEFINE_NOEXCEPT_JIT_OPERATION):
* Source/JavaScriptCore/b3/B3Operations.h: Added.
* Source/JavaScriptCore/b3/B3ReduceStrength.cpp:
* Source/JavaScriptCore/b3/B3Validate.cpp:
* Source/JavaScriptCore/b3/B3Value.cpp:
(JSC::B3::Value::effects const):
* Source/JavaScriptCore/b3/B3Value.h:
* Source/JavaScriptCore/b3/B3ValueInlines.h:
* Source/JavaScriptCore/b3/testb3.h:
* Source/JavaScriptCore/b3/testb3_1.cpp:
(run):
* Source/JavaScriptCore/b3/testb3_8.cpp:
(testMemoryCopy):
(testMemoryCopyConstant):
(testMemoryFill):
(testMemoryFillConstant):
* Source/JavaScriptCore/wasm/WasmOMGIRGenerator.cpp:
(JSC::Wasm::OMGIRGenerator::addMemoryFill):
(JSC::Wasm::OMGIRGenerator::addMemoryCopy):

Canonical link: https://commits.webkit.org/300210@main
